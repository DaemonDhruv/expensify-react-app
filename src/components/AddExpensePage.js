import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        // Redirect programatically after successful submit-dispatch to dashboard
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                Add you expense here
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

// By doing this, we are separating redux-react even apart
// And making it more test friendly
// Now the react component don't have to explicitly use 
// redux methods. As all data and methods will be passed 
// in as normal props

// Why we separated concerns for testing purposes?
// Becuase dispatch() need addExpense() and passing that would
// be difficult and out of our scope here. We only are concerned
// with the working of onSubmit. Whether it is getting called
// with proper values or not.
const matchDispatchToProps = (dispatch) => ({
        // This is going be passed on in the props
        startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, matchDispatchToProps)(AddExpensePage);