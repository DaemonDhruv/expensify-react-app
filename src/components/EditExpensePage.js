import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from '../components/ExpenseForm';
import { editExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        console.log('updated', expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense.id);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onRemove}
                >
                    Remove
                 </button>
            </div>
        );
    }
}

// Here we are searching the id we got from the URL from the router, with the id in the 
// expenses list from the state in the redux store
// And we are extract that particular expense and passing it in the props to our 
// EditExpensePage component
const mapStateToProps = (state, props /* this is the props we got from the router when it called the EditExpensePage component */) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

// mapDispatchToProps = (dispatch, [ownProps]) => ({})
const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: (id) => dispatch(startRemoveExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);