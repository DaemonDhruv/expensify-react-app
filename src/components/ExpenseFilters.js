import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export class ExpenseFilters extends React.Component {
    state = {
        calendarFocused: null
    }


    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }


    onFocusChange = (calendarFocused) => {
        this.setState(() => ({
            calendarFocused
        }))
    }


    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value)
    }


    onSortChange = (e) => {
        //There is a weird bug here... dispatch is so fast.. it
        if (e.target.value === 'date') {
            this.props.sortByDate();
        } else if (e.target.value === 'amount') {
            this.props.sortByAmount();
        }
    }


    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                            className="text-input" 
                            placeholder="Search expenses" 
                            type="text" 
                            value={this.props.filters.text} 
                            onChange={this.onTextChange} />
                    </div>


                    <div className="input-group__item">  
                        <select className="select" value={this.props.filters.sortBy} onChange={this.onSortChange}>
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>


                    <div className="input-group__item">                       
                        {/* Learn more at: https://github.com/airbnb/react-dates#daterangepicker */}
                        <DateRangePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            onDatesChange={this.onDatesChange}
                            focusedInput={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            showClearDates={true}
                            numberOfMonths={1}
                            isOutsideRange={() => false} />
                    </div>
                </div>
            </div>
        );
    }
};


const mapStateToProps = (state) => ({
    filters: state.filters
}); // Along with filters, dispatch() method also gets passed to ExpenseFilters component


const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount())
});


// connect() is working as a middleware between parent component and child component. 
// connect() is injecting data from in between to the child component in it's props
// connect() is also providing methods in the props which will take away data and put into the store

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);