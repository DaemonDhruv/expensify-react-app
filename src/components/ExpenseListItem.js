import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses';
import { Link } from 'react-router-dom';

export const ExpenseListItem = ( { id, count, description, amount, createdAt, dispatch } ) => (
    <div>
        <Link to={`/edit/${id}`}><p>{count}. {description}, {amount} - {createdAt}</p></Link>
        <button
            onClick={ () => {
                dispatch(removeExpense(id))
            }}
        >
            Remove
        </button>
    </div>
); 

export default connect()(ExpenseListItem);