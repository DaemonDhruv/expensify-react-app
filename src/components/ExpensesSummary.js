import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';

export const ExpensesSummary = ({count, total}) => (
    <div>
        <h1>
            Viewing {count} expense{ (count !== 1) && 's' } totalling {numeral(total / 100).format('$0,0.00')}
        </h1>
    </div>
);

const mapStateToProps = (state) => {
    const expenses = getVisibleExpenses(state.expenses, state.filters)
    return {
        count: expenses.length,
        total: getExpensesTotal(expenses)
    }
};

export default connect(mapStateToProps)(ExpensesSummary);
