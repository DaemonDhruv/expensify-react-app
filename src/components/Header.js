import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({startLogout}) => (
    <div>
        <header>
            <h1>Expensify</h1>
            <NavLink to="/dashboard" activeClassName="active-nav-link" >Dashboard</NavLink> 
            <NavLink to="/create" activeClassName="active-nav-link" >Add Expences</NavLink>
            {/* <NavLink to="/help" activeClassName="active-nav-link">Help</NavLink>  */}
            <button onClick={startLogout}>Logout</button>
        </header>
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})

export default connect(undefined, mapDispatchToProps)(Header);