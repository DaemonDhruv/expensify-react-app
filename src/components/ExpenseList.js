import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses';

// exporting it separately for testing purposes
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ?
                <p>Expense list is empty</p>
                :
                (
                    props.expenses.map((expense, index) =>
                        <ExpenseListItem
                            key={expense.id}
                            count={index + 1}
                            {...expense}
                        />)
                )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList);