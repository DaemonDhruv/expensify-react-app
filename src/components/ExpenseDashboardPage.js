import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseFilters from './ExpenseFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpensesSummary />
        <ExpenseFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;