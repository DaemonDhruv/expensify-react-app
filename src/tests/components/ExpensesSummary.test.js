import React from 'react';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import { shallow } from 'enzyme';

test('should address as singular for single expense', () => {
    const wrapper = shallow(<ExpensesSummary count={1} total={94} />);
    expect(wrapper).toMatchSnapshot();
})

test('should address as plural for mulitple expenses', () => {
    const wrapper = shallow(<ExpensesSummary count={2} total={94} />);
    expect(wrapper).toMatchSnapshot();
})