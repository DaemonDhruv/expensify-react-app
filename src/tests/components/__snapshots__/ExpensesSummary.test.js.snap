// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should address as plural for mulitple expenses 1`] = `
<div>
  <h1>
    Viewing 
    2
     expense
    s
     totalling 
    $0.94
  </h1>
</div>
`;

exports[`should address as singular for single expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     expense
     totalling 
    $0.94
  </h1>
</div>
`;
